File: bindings/ruby/test/test_conformance.rb
# frozen_string_literal: true

require_relative "helper"

# The shared 70-case conformance suite: render every case in
# conformance/cases.json and compare byte-for-byte against the goldens in
# conformance/golden/<name>.txt — the contract every ccharts binding is held to.
#
# A settings/struct mismatch in the Fiddle layer shows up here as a byte
# difference, not a crash, so this is the acceptance gate for the binding.

def color(value)
  value # a color name string like "blue", or nil
end

def floats(dataset, key)
  dataset[key].map { |v| v.to_f }
end

def render_case(cases, case_data)
  cfg = case_data["settings"] || {}
  case case_data["chart"]
  when "hist"
    samples = case_data["samples"].map { |v| v.to_f }
    s = Ccharts::Settings::HistSettings.new
    s.bin_count((cfg["bin_count"] || 0).to_i)
    s.min_value(cfg["min_value"].nil? ? Float::NAN : cfg["min_value"].to_f)
    s.max_value(cfg["max_value"].nil? ? Float::NAN : cfg["max_value"].to_f)
    s.show_bins(cfg["show_bins"] || false)
    s.show_prices(cfg["show_prices"] || false)
    s.plain(cfg["plain"] || false)
    s.rise(color(cfg["rise_color"])) if cfg["rise_color"]
    s.background(color(cfg["bg_color"])) if cfg["bg_color"]
    Ccharts::Chart.histogram(samples, case_data["width"], case_data["height"], s)

  when "spark"
    samples = case_data["samples"].map { |v| v.to_f }
    s = Ccharts::Settings::SparkSettings.new
    s.min_above((cfg["min_above"] || 0).to_i)
    s.min_below((cfg["min_below"] || 0).to_i)
    s.plain(cfg["plain"] || false)
    s.rise(color(cfg["rise_color"])) if cfg["rise_color"]
    s.area(color(cfg["area_color"])) if cfg["area_color"]
    Ccharts::Chart.sparkline(samples, case_data["width"], case_data["height"], s)

  when "bar"
    labels = case_data["items"].map { |it| it["label"].to_s }
    values = case_data["items"].map { |it| it["value"].to_f }
    s = Ccharts::Settings::BarSettings.new
    s.show_labels(cfg["show_labels"] || false)
    s.show_prices(cfg["show_prices"] || false)
    s.plain(cfg["plain"] || false)
    s.rise(color(cfg["rise_color"])) if cfg["rise_color"]
    s.background(color(cfg["bg_color"])) if cfg["bg_color"]
    Ccharts::Chart.bar(labels, values, case_data["width"], case_data["height"], s)

  when "pie"
    slices = case_data["slices"].map { |sl| { label: sl["label"].to_s, value: sl["value"].to_f } }
    s = Ccharts::Settings::PieSettings.new
    s.donut(cfg["donut"] || false)
    s.show_legend(cfg["show_legend"].nil? ? true : cfg["show_legend"])
    s.show_pct(cfg["show_pct"] || false)
    s.slice_gap(cfg["slice_gap"].to_f) if cfg["slice_gap"]
    s.inner_radius_ratio(cfg["inner_radius_ratio"].to_f) if cfg["inner_radius_ratio"]
    s.legend_format(cfg["legend_format"].to_i) if cfg["legend_format"]
    s.start_angle(cfg["start_angle"].to_f) if cfg["start_angle"]
    s.counter_clockwise(cfg["counter_clockwise"] || false)
    s.center_text(cfg["center_text"]) if cfg["center_text"]
    s.colors(cfg["colors"]) if cfg["colors"]
    Ccharts::Chart.pie(slices, case_data["width"], case_data["height"], s)

  when "stack"
    series = case_data["series"].map do |sr|
      { name: sr["name"].to_s, values: sr["values"].map { |v| v.to_f } }
    end
    s = Ccharts::Settings::StackSettings.new
    s.show_labels(cfg["show_labels"] || false)
    s.show_prices(cfg["show_prices"] || false)
    s.plain(cfg["plain"] || false)
    s.colors(cfg["colors"]) if cfg["colors"]
    s.background(color(cfg["bg_color"])) if cfg["bg_color"]
    s.category_labels(cfg["cat_labels"]) if cfg["cat_labels"]
    Ccharts::Chart.stacked_bar(series, case_data["width"], case_data["height"], s)

  when "heat"
    matrix = case_data["values"].map { |row| row.map { |v| v.to_f } }
    s = Ccharts::Settings::HeatSettings.new
    s.show_labels(cfg["show_labels"] || false)
    s.plain(cfg["plain"] || false)
    s.low_color(color(cfg["low_color"])) if cfg["low_color"]
    s.high_color(color(cfg["high_color"])) if cfg["high_color"]
    s.mid_color(color(cfg["mid_color"])) if cfg["mid_color"]
    s.background(color(cfg["bg_color"])) if cfg["bg_color"]
    s.row_labels(cfg["row_labels"]) if cfg["row_labels"]
    s.col_labels(cfg["col_labels"]) if cfg["col_labels"]
    Ccharts::Chart.heatmap(matrix, case_data["width"], case_data["height"], s)

  when "box"
    categories = case_data["categories"].map do |c|
      { name: c["name"].to_s, samples: c["samples"].map { |v| v.to_f } }
    end
    s = Ccharts::Settings::BoxSettings.new
    s.show_prices(cfg["show_prices"] || false)
    s.plain(cfg["plain"] || false)
    s.rise(color(cfg["rise_color"])) if cfg["rise_color"]
    s.area(color(cfg["area_color"])) if cfg["area_color"]
    s.background(color(cfg["bg_color"])) if cfg["bg_color"]
    Ccharts::Chart.boxplot(categories, case_data["width"], case_data["height"], s)

  else # line / candle
    dataset = cases["datasets"][case_data["dataset"]]
    chart = if case_data["source"] == "json"
              Ccharts::Chart.from_json(dataset["json"])
            else
              ts = dataset["ts"] ? dataset["ts"].map(&:to_i) : nil
              Ccharts::Chart.from_arrays(
                open: floats(dataset, "open"), high: floats(dataset, "high"),
                low: floats(dataset, "low"), close: floats(dataset, "close"), ts: ts
              )
            end
    s = Ccharts::Settings::ChartSettings.new
    s.single_color(cfg["single_color"] || false)
    s.show_prices(cfg["show_prices"] || false)
    s.show_times(cfg["show_times"] || false)
    s.plain(cfg["plain"] || false)
    s.rise(color(cfg["rise_color"])) if cfg["rise_color"]
    s.fall(color(cfg["fall_color"])) if cfg["fall_color"]
    s.background(color(cfg["bg_color"])) if cfg["bg_color"]
    s.area(color(cfg["area_color"])) if cfg["area_color"]
    if case_data["chart"] == "line"
      chart.line(case_data["width"], case_data["height"], s)
    else
      chart.candle(case_data["width"], case_data["height"], s)
    end
  end
end

test "all 70 conformance cases match the goldens byte-for-byte" do
  dir = conformance_dir
  raise "conformance suite not present at #{dir}" unless File.exist?(File.join(dir, "cases.json"))

  doc = JSON.parse(File.read(File.join(dir, "cases.json")))
  cases = doc["cases"]
  assert cases.length >= 70, "conformance suite looks truncated (#{cases.length} cases)"

  failures = []
  cases.each do |case_data|
    name = case_data["name"]
    rendered = render_case(doc, case_data)
    expected_path = File.join(dir, "golden", "#{name}.txt")
    raise "missing golden file for #{name}" unless File.exist?(expected_path)
    expected = File.binread(expected_path)
    unless rendered.b == expected
      failures << name
    end
  end

  unless failures.empty?
    raise "#{failures.length} case(s) differ from the goldens: #{failures.inspect}"
  end
  assert true
end

TinyTest.run

File: bindings/julia/test/test_conformance.jl
# test/test_conformance.jl — the shared 70-case conformance suite.
#
# Every case in ../../conformance/cases.json is rendered through the public
# Ccharts API and compared byte-for-byte against ../../conformance/golden/<name>.txt.
# A settings/struct mismatch shows up as a byte difference, so this is the
# acceptance gate for the binding.
using Test
using Ccharts
using JSON

CONF_DIR = normpath(joinpath(@__DIR__, "..", "..", "..", "conformance"))

# cfg field accessor that treats JSON null (Julia `nothing`) as the default.
_cfg(cfg, k, dflt) = haskey(cfg, k) && cfg[k] !== nothing ? cfg[k] : dflt

function _floats(v)
    v === nothing ? nothing : Float64.(v)
end

function render_case(doc, case_data)
    cfg = get(case_data, "settings", Dict())
    chart = case_data["chart"]
    w = case_data["width"]; h = case_data["height"]

    if chart == "hist"
        samples = Float64.(case_data["samples"])
        return histogram(samples, w, h;
            rise_color=_cfg(cfg,"rise_color",nothing),
            bg_color=_cfg(cfg,"bg_color",nothing),
            bin_count=_cfg(cfg,"bin_count",0),
            min_value=_cfg(cfg,"min_value",NaN),
            max_value=_cfg(cfg,"max_value",NaN),
            show_bins=_cfg(cfg,"show_bins",false),
            show_prices=_cfg(cfg,"show_prices",false),
            plain=_cfg(cfg,"plain",false))

    elseif chart == "spark"
        samples = Float64.(case_data["samples"])
        return sparkline(samples, w, h;
            rise_color=_cfg(cfg,"rise_color",nothing),
            area_color=_cfg(cfg,"area_color",nothing),
            min_above=_cfg(cfg,"min_above",0),
            min_below=_cfg(cfg,"min_below",0),
            plain=_cfg(cfg,"plain",false))

    elseif chart == "bar"
        labels = [string(it["label"]) for it in case_data["items"]]
        values = Float64.([it["value"] for it in case_data["items"]])
        return bar(collect(zip(labels, values)), w, h;
            rise_color=_cfg(cfg,"rise_color",nothing),
            bg_color=_cfg(cfg,"bg_color",nothing),
            show_labels=_cfg(cfg,"show_labels",false),
            show_prices=_cfg(cfg,"show_prices",false),
            plain=_cfg(cfg,"plain",false))

    elseif chart == "pie"
        slices = [(string(sl["label"]), Float64(sl["value"])) for sl in case_data["slices"]]
        return pie(slices, w, h;
            donut=_cfg(cfg,"donut",false),
            colors=_cfg(cfg,"colors",nothing),
            show_legend=_cfg(cfg,"show_legend",true),
            show_pct=_cfg(cfg,"show_pct",false),
            slice_gap=_cfg(cfg,"slice_gap",0.0),
            inner_radius_ratio=_cfg(cfg,"inner_radius_ratio",-1.0),
            legend_format=_cfg(cfg,"legend_format",0),
            start_angle=_cfg(cfg,"start_angle",-1.0),
            counter_clockwise=_cfg(cfg,"counter_clockwise",false),
            center_text=_cfg(cfg,"center_text",nothing))

    elseif chart == "stack"
        series = [(string(sr["name"]), Float64.(sr["values"])) for sr in case_data["series"]]
        return stacked_bar(series, w, h;
            colors=_cfg(cfg,"colors",nothing),
            bg_color=_cfg(cfg,"bg_color",nothing),
            cat_labels=_cfg(cfg,"cat_labels",nothing),
            show_labels=_cfg(cfg,"show_labels",false),
            show_prices=_cfg(cfg,"show_prices",false),
            plain=_cfg(cfg,"plain",false))

    elseif chart == "heat"
        matrix = [Float64.(row) for row in case_data["values"]]
        return heatmap(matrix, w, h;
            low_color=_cfg(cfg,"low_color",nothing),
            high_color=_cfg(cfg,"high_color",nothing),
            mid_color=_cfg(cfg,"mid_color",nothing),
            bg_color=_cfg(cfg,"bg_color",nothing),
            row_labels=_cfg(cfg,"row_labels",nothing),
            col_labels=_cfg(cfg,"col_labels",nothing),
            show_labels=_cfg(cfg,"show_labels",false),
            plain=_cfg(cfg,"plain",false))

    elseif chart == "box"
        categories = [(string(c["name"]), Float64.(c["samples"])) for c in case_data["categories"]]
        return boxplot(categories, w, h;
            rise_color=_cfg(cfg,"rise_color",nothing),
            area_color=_cfg(cfg,"area_color",nothing),
            bg_color=_cfg(cfg,"bg_color",nothing),
            show_prices=_cfg(cfg,"show_prices",false),
            plain=_cfg(cfg,"plain",false))

    else # line / candle
        dataset = doc["datasets"][case_data["dataset"]]
        c = if get(case_data,"source","arrays") == "json"
            Chart.from_json(dataset["json"])
        else
            ts = isempty(dataset) || !haskey(dataset,"ts") ? nothing : Int64[round(Int,x) for x in dataset["ts"]]
            Chart.from_arrays(_floats(dataset["open"]), _floats(dataset["high"]),
                              _floats(dataset["low"]), _floats(dataset["close"]); ts=ts)
        end
        return (chart == "line" ? line : candle)(c, w, h;
            rise_color=_cfg(cfg,"rise_color",nothing),
            fall_color=_cfg(cfg,"fall_color",nothing),
            bg_color=_cfg(cfg,"bg_color",nothing),
            area_color=_cfg(cfg,"area_color",nothing),
            single_color=_cfg(cfg,"single_color",false),
            show_prices=_cfg(cfg,"show_prices",false),
            show_times=_cfg(cfg,"show_times",false),
            plain=_cfg(cfg,"plain",false))
    end
end

@testset "all 70 conformance cases match the goldens byte-for-byte" begin
    cases_file = joinpath(CONF_DIR, "cases.json")
    @test isfile(cases_file)
    doc = JSON.parsefile(cases_file)
    cases = doc["cases"]
    @test length(cases) >= 70  # truncation guard

    failures = String[]
    for case_data in cases
        name = case_data["name"]
        rendered = render_case(doc, case_data)
        expected_path = joinpath(CONF_DIR, "golden", name * ".txt")
        @test isfile(expected_path)
        expected = read(expected_path, String)
        rendered == expected || push!(failures, name)
    end
    @test isempty(failures)
    if !isempty(failures)
        println("conformance mismatches: ", failures)
    end
    @test length(cases) == 70
end